magefile.go

```diff
@@ -1174,9 +1179,248 @@ func collectPackageDependencies(platforms []string, packageVersion string, requi
+			log.Printf("> prepare to copy %s into %s ", f, versionedDropPath)
+		}
+
+		err = copy.Copy(f, versionedDropPath, options)
+		if err != nil {
+			panic(err)
+		}
+
+		// copy spec file for match
+		specName := filepath.Base(f)
+		idx := strings.Index(specName, "-"+packageVersion)
+		if idx != -1 {
+			specName = specName[:idx]
+		}
+		if mg.Verbose() {
+			log.Printf(">>>> Looking to copy spec file: [%s]", specName)
+		}
+
+		checksum, err := copyComponentSpecs(specName, versionedDropPath)
+		if err != nil {
+			panic(err)
+		}
+
+		checksums[specName+specSuffix] = checksum
+	}
+
+	return checksums
+}
+
+// This function is used when building with a Manifest.  In that manifest, it's possible
+// for projects in an Independent Agent Release to have different versions since the opted-in
+// ones will be one patch version higher than the opted-out/previously released projects.
+// This function tries to find the versions from the package name
+func getComponentVersion(componentName string, requiredPackage string, componentProject tools.Project) string {
+	var componentVersion string
+	var foundIt bool
+	// Iterate over all the packages in the component project
+	for pkgName, _ := range componentProject.Packages {
+		// Only care about the external binaries that we want to package
+		for binaryPrefix, binaryComponent := range externalBinaries {
+			// If the given component name doesn't match the external binary component, skip
+			if componentName != binaryComponent {
+				continue
+			}
+
+			// Split the package name on the binary name prefix plus a dash
+			firstSplit := strings.Split(pkgName, binaryPrefix+"-")
+			if len(firstSplit) < 2 {
+				continue
+			}
+
+			// Get the second part of the first split
+			secondHalf := firstSplit[1]
+			if len(secondHalf) < 2 {
+				continue
+			}
+
+			// Make sure the second half matches the required package
+			if strings.Contains(secondHalf, requiredPackage) {
+				// ignore packages with names where this splitting doesn't results in proper version
+				if strings.Contains(secondHalf, "docker-image") {
+					continue
+				}
+				if strings.Contains(secondHalf, "oss-") {
+					continue
+				}
+
+				// The component version should be the first entry after splitting w/ the requiredPackage
+				componentVersion = strings.Split(secondHalf, "-"+requiredPackage)[0]
+				foundIt = true
+				// break out of inner loop
+				break
+			}
+		}
+		if foundIt {
+			// break out of outer loop
+			break
+		}
+	}
+
+	if componentVersion == "" {
+		errMsg := fmt.Sprintf("Unable to determine component version for [%s]", componentName)
+		panic(errMsg)
+	}
+
+	return componentVersion
+}
+
+// This is a helper function for the cloud-defend package.
+// When it is untarred, it does not have the same dirname as the package name.
+// This adjusts for that and returns the actual path on disk for cloud-defend
+func fixCloudDefendDirPath(dirPath string, componentVersion string, expectedArch string, actualArch string) string {
```

Comment:
Could we just fix this in cloud-defend directly at some point? https://github.com/elastic/cloud-defend

Reply:
Also FYI that we only include defend in the container build: https://github.com/elastic/elastic-agent/pull/4584/files